ddl/multi_schema_change.go

```diff
@@ -56,18 +57,137 @@ func (d *ddl) MultiSchemaChange(ctx sessionctx.Context, ti ast.Ident) error {
+	if job.MultiSchemaInfo.Revertible {
+		// Handle the rolling back job.
+		if job.IsRollingback() {
+			// Rollback/cancel the sub-jobs in reverse order.
+			for i := len(job.MultiSchemaInfo.SubJobs) - 1; i >= 0; i-- {
+				sub := job.MultiSchemaInfo.SubJobs[i]
+				if sub.IsFinished() {
+					continue
+				}
+				proxyJob := sub.ToProxyJob(job)
+				ver, err = w.runDDLJob(d, t, proxyJob)
+				sub.FromProxyJob(proxyJob)
+				if i == 0 && sub.IsFinished() {
+					job.State = model.JobStateRollbackDone
+				}
+				return ver, err
+			}
+			// The last rollback/cancelling sub-job is done.
+			job.State = model.JobStateRollbackDone
+			return ver, nil
+		}
+
+		// The sub-jobs are normally running.
+		// Run the first executable sub-job.
+		for _, sub := range job.MultiSchemaInfo.SubJobs {
+			if !sub.Revertible || sub.IsFinished() {
+				// Skip the sub-jobs which related schema states
+				// are in the last revertible point.
+				// If a sub job is finished here, it should be a noop job.
+				continue
+			}
+			proxyJob := sub.ToProxyJob(job)
+			ver, err = w.runDDLJob(d, t, proxyJob)
+			sub.FromProxyJob(proxyJob)
+			handleRevertibleException(job, sub, proxyJob.Error)
+			return ver, err
+		}
+
+		// Save tblInfo and subJobs for rollback, because some DDLs update the transaction aggressively.
+		tblInfo, _ := t.GetTable(job.SchemaID, job.TableID)
+		subJobs := make([]model.SubJob, len(job.MultiSchemaInfo.SubJobs))
+		// Step the sub-jobs to the non-revertible states all at once.
+		for i, sub := range job.MultiSchemaInfo.SubJobs {
+			if sub.IsFinished() {
+				continue
+			}
+			subJobs[i] = *sub
+			proxyJob := sub.ToProxyJob(job)
+			ver, err = w.runDDLJob(d, t, proxyJob)
+			sub.FromProxyJob(proxyJob)
+			if err != nil || proxyJob.Error != nil {
+				for j := i - 1; j >= 0; j-- {
+					job.MultiSchemaInfo.SubJobs[j] = &subJobs[j]
+				}
+				handleRevertibleException(job, sub, proxyJob.Error)
+				return updateVersionAndTableInfo(d, t, job, tblInfo, true)
+			}
+		}
+		// All the sub-jobs are non-revertible.
+		job.MultiSchemaInfo.Revertible = false
+		return ver, err
+	}
+	// Run the rest non-revertible sub-jobs one by one.
+	for _, sub := range job.MultiSchemaInfo.SubJobs {
+		if sub.IsFinished() {
+			continue
+		}
+		proxyJob := sub.ToProxyJob(job)
+		ver, err = w.runDDLJob(d, t, proxyJob)
+		sub.FromProxyJob(proxyJob)
+		return ver, err
 	}
+	job.State = model.JobStateDone
+	return ver, err
+}
 
-	err = checkVisibleColumnCnt(t, len(info.AddColumns), len(info.DropColumns))
+func handleRevertibleException(job *model.Job, subJob *model.SubJob, err *terror.Error) {
+	if subJob.IsNormal() {
+		return
+	}
+	job.State = model.JobStateRollingback
+	job.Error = err
+	// Flush the cancelling state and cancelled state to sub-jobs.
+	for _, sub := range job.MultiSchemaInfo.SubJobs {
+		switch sub.State {
+		case model.JobStateCancelled:
+			if !sub.Revertible {
+				sub.State = model.JobStateCancelling
+			}
```

Comment:
Why do this？

Reply:
please add comment for this.

Reply:
It is for `onModifyColumn`. Let me remove it in this PR.